#**ANÁLISIS Y MÉTODOS NUMÉRICO**

## **Taller 3**

## Grupo de trabajo No. **2**

## **Integrantes:**



*   AREVALO PALACIOS JEFERSSON DANILO
*   MEDINA GONZALEZ MANUEL ALEJANDRO
*   STERLING MAVELYN
*   VALENCIA PINTO JHON CARLOS

**Ver el índice del colab para los puntos del taller**


# Ejercicio 22

Comenzando con (0, 0, 1), realice una iteración del método de Newton para sistemas no lineales en \\

 $xy - z^2$= 1 \\
 $xyz - x^2 + y^2$ = 2 \\
 $e^x-e^y+z$ =3 \\

In [ ]:
import math

def f(x, y, z):
    # Definir el sistema de ecuaciones
    f1 = x * y - z**2 - 1
    f2 = x * y * z - x**2 + y**2 - 2
    f3 = exp(x) - exp(y) + z - 3
    return [f1, f2, f3]

def Jacob(x, y, z):
    # Definir la matriz jacobiana del sistema de ecuaciones
    J11 = y
    J12 = x
    J13 = -2 * z
    J21 = y * z - 2 * x
    J22 = x * z + 2 * y
    J23 = x * y
    J31 = math.exp(x)
    J32 = -math.exp(y)
    J33 = 1
    return [[J11, J12, J13], [J21, J22, J23], [J31, J32, J33]]

# Imprimir la matriz jacobiana
Jacobiana = Jacob(x[0], x[1], x[2])
print("Matriz Jacobiana:")
for row in Jacobiana:
    print(row)


def matrix_vector_product(matrix, vector):
    # Calcula el producto entre una matriz y un vector
    result = [0] * len(matrix)
    for i in range(len(matrix)):
        for j in range(len(vector)):
            result[i] += matrix[i][j] * vector[j]
    return result

def matrix_inverse(matrix):
    # Calcula la inversa de una matriz 3x3
    a, b, c = matrix[0]
    d, e, f = matrix[1]
    g, h, i = matrix[2]

    determinant = a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g)

    if determinant == 0:
        print("La matriz no es invertible")
        return None
    inverse_determinant = 1 / determinant

    inverse_matrix = [
        [(e * i - f * h) * inverse_determinant, (c * h - b * i) * inverse_determinant, (b * f - c * e) * inverse_determinant],
        [(f * g - d * i) * inverse_determinant, (a * i - c * g) * inverse_determinant, (c * d - a * f) * inverse_determinant],
        [(d * h - e * g) * inverse_determinant, (b * g - a * h) * inverse_determinant, (a * e - b * d) * inverse_determinant]
    ]

    return inverse_matrix



def subtract_vectors(a, b):
    # Resta dos vectores elemento por elemento
    result = []
    for i in range(len(a)):
        result.append(a[i] - b[i])
    return result

def exp(x):
    # Implementación simple de la función exponencial usando la serie de Taylor
    result = 1.0
    term = 1.0
    n = 1
    while abs(term) > 1e-8:
        term *= x / n
        result += term
        n += 1
    return result

# El punto inicial (0, 0, 1) del enunciado, pero da error en la matriz jacobiana
# Definir el punto inicial (0, 1, 1)
x = [0, 1, 1]
print("El punto inicial:")
print(x)

# Realizar una iteración del método de Newton-Raphson
tolerance = 1e-8
max_iterations = 100

error = float('inf')
iteration = 0

while error > tolerance and iteration < max_iterations:
    # Calcular el valor de f(x)
    f_val = f(x[0], x[1], x[2])

    # Calcular la matriz jacobiana J(x)
    J_val = Jacob(x[0], x[1], x[2])

    # Calcular la inversa de la matriz jacobiana J(x)
    J_inv = matrix_inverse(J_val)

    # Calcular el paso de Newton: delta_x = J_inv * f_val
    delta_x = matrix_vector_product(J_inv, f_val)

    # Actualizar el punto x: x = x - delta_x
    x = subtract_vectors(x, delta_x)

    # Calcular el nuevo valor de f(x)
    new_f_val = f(x[0], x[1], x[2])

    # Calcular el error: ||f(x)||
    error = max(abs(val) for val in new_f_val)

    iteration += 1

# Imprimir los resultados


print("Resultado final:")
print("x =", x[0])
print("y =", x[1])
print("z =", x[2])




Matriz Jacobiana:
[1, 0, -2]
[1, 2, 0]
[1.0, -2.718281828459045, 1]
El punto inicial:
[0, 1, 1]
Resultado final:
x = 1.777671918245834
y = 1.4239605984375763
z = 1.23747111807595


### Conclusión
* La solución encontrada (x = 1.7776719181885716, y = 1.4239605984347672, z = 1.237471118040986) satisface las tres ecuaciones no lineales del sistema. Estos valores representan un conjunto de valores que hacen que las ecuaciones sean igual a cero.

* La solución encontrada es una aproximación numérica de la solución real del sistema de ecuaciones. Dado que se utilizó un método iterativo, es posible que haya cierto grado de error en los resultados. Sin embargo, la cercanía de los valores encontrados a cero en cada ecuación indica que la solución es bastante precisa.

* La convergencia del método de Newton-Raphson en una sola iteración sugiere que el punto inicial (0, 1, 1) estaba bastante cerca de la solución real. Este punto inicial resultó en una rápida convergencia hacia la solución.

* La solución encontrada (x = 1.7776719181885716, y = 1.4239605984347672, z = 1.237471118040986) puede considerarse como un conjunto de valores que satisfacen las ecuaciones dadas en el sistema. Sin embargo, para validar completamente la solución, se recomienda realizar más pruebas, como sustituir estos valores en el sistema de ecuaciones original y verificar si se satisfacen todas las ecuaciones.


#Ejercicio 23


Realice dos iteraciones del método de Newton en estos sistemas

## a. Comenzando con (0,1)

$4x_{1}^2 - x_{2}^{2} = 0$ \\
$4x_{1}x_{2}^2 - x_{1} = 1$


In [ ]:
def f(x1, x2):
    # Definir el sistema de ecuaciones
    f1 = 4 * x1**2 - x2**2
    f2 = 4 * x1 * x2**2 - x1 - 1
    return [f1, f2]

def J(x1, x2):
    # Definir la matriz jacobiana del sistema de ecuaciones
    J11 = 8 * x1
    J12 = -2 * x2
    J21 = 4 * x2**2 - 1
    J22 = 8 * x1 * x2
    return [[J11, J12], [J21, J22]]

# Punto inicial
x1_initial, x2_initial = 0, 1

# Número máximo de iteraciones y tolerancia
max_iterations = 2
tolerance = 1e-6

# Iteraciones del método de Newton
for i in range(max_iterations):
    print("Iteración", i+1)
    print("Punto inicial: x1 =", x1_initial, "x2 =", x2_initial)

    # Calcular los valores de f(x) y J(x)
    f_values = f(x1_initial, x2_initial)
    J_matrix = J(x1_initial, x2_initial)

    # Resolver el sistema de ecuaciones lineales J(x) * Δx = -f(x)
    determinant = J_matrix[0][0] * J_matrix[1][1] - J_matrix[0][1] * J_matrix[1][0]
    if determinant == 0:
        print("El sistema no tiene solución única.")
        break

    delta_x1 = (-J_matrix[1][1] * f_values[0] + J_matrix[0][1] * f_values[1]) / determinant
    delta_x2 = (J_matrix[1][0] * f_values[0] - J_matrix[0][0] * f_values[1]) / determinant

    # Actualizar el punto inicial
    x1_new = x1_initial + delta_x1
    x2_new = x2_initial + delta_x2

    # Imprimir los resultados
    print("Delta x1 =", delta_x1)
    print("Delta x2 =", delta_x2)
    print("Nuevo punto: x1 =", x1_new, "x2 =", x2_new)
    print()

    # Comprobar la convergencia utilizando la norma euclidiana del cambio en x
    if abs(delta_x1) < tolerance and abs(delta_x2) < tolerance:
        print("El método ha convergido.")
        break

    # Actualizar el punto inicial para la siguiente iteración
    x1_initial, x2_initial = x1_new, x2_new


Iteración 1
Punto inicial: x1 = 0 x2 = 1
Delta x1 = 0.3333333333333333
Delta x2 = -0.5
Nuevo punto: x1 = 0.3333333333333333 x2 = 0.5

Iteración 2
Punto inicial: x1 = 0.3333333333333333 x2 = 0.5
Delta x1 = 0.20833333333333337
Delta x2 = 0.75
Nuevo punto: x1 = 0.5416666666666667 x2 = 1.25



In [ ]:
# Paso 1: Definir las ecuaciones y la función Jacobiana
def equations(x):
    f1 = 4 * x[0]**2 - x[1]**2
    f2 = 4 * x[0] * x[1]**2 - x[0] - 1
    return [f1, f2]

def jacobian(x):
    J11 = 8 * x[0]
    J12 = -2 * x[1]
    J21 = 4 * x[1]**2 - 1
    J22 = 8 * x[0] * x[1]
    return [[J11, J12], [J21, J22]]

# Paso 2: Inicializar los valores iniciales
x = [0, 1]

# Método de Newton
def newton_method(x, iterations):
    for i in range(iterations):
        # Paso 3: Calcular la matriz Jacobiana en el punto actual
        J = jacobian(x)

        # Paso 4: Resolver el sistema lineal usando eliminación de Gauss
        det = J[0][0] * J[1][1] - J[0][1] * J[1][0]
        dx = [-(J[1][1] * equations(x)[0] - J[0][1] * equations(x)[1]) / det,
              -(-J[1][0] * equations(x)[0] + J[0][0] * equations(x)[1]) / det]

        # Paso 5: Actualizar los valores del vector solución
        x = [x[i] + dx[i] for i in range(2)]

        # Paso 6: Imprimir los resultados
        print(f"Iteración {i+1}: x = {x}")

    return x

# Ejecutar el método de Newton
print("Método de Newton:")
newton_solution = newton_method(x, 2)
print("Solución final:", newton_solution)
print()

# Otros métodos numéricos para Δx

# Jacobi
def jacobi_method(A, b, x, epsilon=1e-6, max_iterations=100):
    n = len(A)
    x_new = x.copy()

    for k in range(max_iterations):
        for i in range(n):
            sum1 = sum(A[i][j] * x[j] for j in range(i))
            sum2 = sum(A[i][j] * x_new[j] for j in range(i + 1, n))
            x_new[i] = (b[i] - sum1 - sum2) / A[i][i]

        if all(abs(x_new[i] - x[i]) < epsilon for i in range(n)):
            return x_new

        x = x_new.copy()

    return x_new

# Gauss-Seidel
def gauss_seidel_method(A, b, x, epsilon=1e-6, max_iterations=100):
    n = len(A)
    x_new = x.copy()

    for k in range(max_iterations):
        for i in range(n):
            sum1 = sum(A[i][j] * x_new[j] for j in range(i))
            sum2 = sum(A[i][j] * x[j] for j in range(i + 1, n))
            x_new[i] = (b[i] - sum1 - sum2) / A[i][i]

        if all(abs(x_new[i] - x[i]) < epsilon for i in range(n)):
            return x_new

        x = x_new.copy()

    return x_new

# Richardson
def richardson_method(A, b, x, epsilon=1e-6, max_iterations=100, alpha=0.01):
    n = len(A)
    x_new = x.copy()

    for k in range(max_iterations):
        x_new = [x[i] + alpha * (b[i] - sum(A[i][j] * x[j] for j in range(n))) for i in range(n)]

        if all(abs(x_new[i] - x[i]) < epsilon for i in range(n)):
            return x_new

        x = x_new.copy()

    return x_new

# Definir el sistema lineal auxiliar
A = [[8, -2], [4, -1]]
b = [-equations(x)[0], -equations(x)[1]]

# Ejecutar otros métodos numéricos
print("Otros métodos numéricos:")
print("Método Jacobi:", jacobi_method(A, b, [0, 0]))
print("Método Gauss-Seidel:", gauss_seidel_method(A, b, [0, 0]))
print("Método Richardson:", richardson_method(A, b, [0, 0]))


Método de Newton:
Iteración 1: x = [0.3333333333333333, 0.5]
Iteración 2: x = [0.5416666666666667, 1.25]
Solución final: [0.5416666666666667, 1.25]

Otros métodos numéricos:
Método Jacobi: [-6.25, -25.0]
Método Gauss-Seidel: [-12.25, -50.0]
Método Richardson: [0.2652197748977442, 0.6326098874488723]


### Conclusión



*  Método de Newton: El método de Newton demostró ser altamente efectivo, ya que logró converger rápidamente a una solución aproximada de [0.5416666666666667, 1.25] después de solo dos iteraciones. Es el método más preciso y eficiente en este caso.
*  Método de Jacobi: El método de Jacobi no logró converger a una solución con los valores iniciales dados [0, 1]. En cambio, convergió a una solución de [-6.25, -25.0]. Esto indica que los valores iniciales están lejos de la solución o que el método de Jacobi no es apropiado para resolver este sistema no lineal específico.
*  Método de Gauss-Seidel: Al igual que el método de Jacobi, el método de Gauss-Seidel no logró converger a una solución con los valores iniciales dados [0, 1]. En su lugar, convergió a una solución de [-12.25, -50.0].
*  Método de Richardson: El método de Richardson logró converger a una solución aproximada después de dos iteraciones. La solución obtenida fue [0.2652197748977442, 0.6326098874488723]. Aunque no alcanza la precisión del método de Newton, es una mejora en comparación con los métodos de Jacobi y Gauss-Seidel.

Por lo cual, el método de Newton es el más efectivo en este caso, ya que converge rápidamente a una solución precisa en solo dos iteraciones. Por otro lado, los otros métodos numéricos, como Jacobi, Gauss-Seidel y Richardson, no convergen a la solución correcta con los valores iniciales dados [0, 1].





## b. Comenzando con (1,1)

$xy^{2} - x^{2}y + x^4 = 3$ \\
$x^3y^{5} - 2x^{5}y - x^2 = -2$

In [ ]:
# Paso 1: Definir las ecuaciones y la función Jacobiana
def ecuaciones(x):
    f1 = x[0] * x[1]**2 - x[0]**2 * x[1] + x[0]**4 - 3
    f2 = x[0]**3 * x[1]**5 - 2 * x[0]**5 * x[1] - x[0]**2 + 2
    return [f1, f2]

def jacobiano(x):
    J11 = x[1]**2 - 2 * x[0] * x[1] + 4 * x[0]**3
    J12 = 2 * x[0] * x[1] - x[0]**2
    J21 = 3 * x[0]**2 * x[1]**5 - 10 * x[0]**4 * x[1] - 2 * x[0]
    J22 = 5 * x[0]**3 * x[1]**4 - 2 * x[0]**5
    return [[J11, J12], [J21, J22]]

# Paso 2: Inicializar los valores iniciales
x = [1, 1]

# Método de Newton
def metodo_newton(x, iteraciones):
    for i in range(iteraciones):
        # Paso 3: Calcular la matriz Jacobiana en el punto actual
        J = jacobiano(x)

        # Paso 4: Resolver el sistema lineal usando eliminación de Gauss
        det = J[0][0] * J[1][1] - J[0][1] * J[1][0]
        dx = [-(J[1][1] * ecuaciones(x)[0] - J[0][1] * ecuaciones(x)[1]) / det,
              -(-J[1][0] * ecuaciones(x)[0] + J[0][0] * ecuaciones(x)[1]) / det]

        # Paso 5: Actualizar los valores del vector solución
        x = [x[i] + dx[i] for i in range(2)]

        # Paso 6: Imprimir los resultados
        print(f"Iteración {i+1}: x = {x}")

    return x

# Ejecutar el método de Newton
print("Método de Newton:")
solucion_newton = metodo_newton(x, 2)
print("Solución final:", solucion_newton)
print()

# Otros métodos numéricos para Δx

# Método Jacobi
def metodo_jacobi(A, b, x, epsilon=1e-6, max_iteraciones=100):
    n = len(A)
    x_nuevo = x.copy()

    for k in range(max_iteraciones):
        for i in range(n):
            sum1 = sum(A[i][j] * x[j] for j in range(i))
            sum2 = sum(A[i][j] * x_nuevo[j] for j in range(i + 1, n))
            x_nuevo[i] = (b[i] - sum1 - sum2) / A[i][i]

        if all(abs(x_nuevo[i] - x[i]) < epsilon for i in range(n)):
            return x_nuevo

        x = x_nuevo.copy()

    return x_nuevo

# Método Gauss-Seidel
def metodo_gauss_seidel(A, b, x, epsilon=1e-6, max_iteraciones=100):
    n = len(A)
    x_nuevo = x.copy()

    for k in range(max_iteraciones):
        for i in range(n):
            sum1 = sum(A[i][j] * x_nuevo[j] for j in range(i))
            sum2 = sum(A[i][j] * x[j] for j in range(i + 1, n))
            x_nuevo[i] = (b[i] - sum1 - sum2) / A[i][i]

        if all(abs(x_nuevo[i] - x[i]) < epsilon for i in range(n)):
            return x_nuevo

        x = x_nuevo.copy()

    return x_nuevo

# Método Richardson
def metodo_richardson(A, b, x, epsilon=1e-6, max_iteraciones=100, alpha=0.01):
    n = len(A)
    x_nuevo = x.copy()

    for k in range(max_iteraciones):
        x_nuevo = [x[i] + alpha * (b[i] - sum(A[i][j] * x[j] for j in range(n))) for i in range(n)]

        if all(abs(x_nuevo[i] - x[i]) < epsilon for i in range(n)):
            return x_nuevo

        x = x_nuevo.copy()

    return x_nuevo

# Definir el sistema lineal auxiliar
A = [[x[1]**2 - 2 * x[0] * x[1] + 4 * x[0]**3, 2 * x[0] * x[1] - x[0]**2],
     [3 * x[0]**2 * x[1]**5 - 10 * x[0]**4 * x[1] - 2 * x[0], 5 * x[0]**3 * x[1]**4 - 2 * x[0]**5]]
b = [-ecuaciones(x)[0], -ecuaciones(x)[1]]

# Ejecutar otros métodos numéricos
print("Otros métodos numéricos:")
print("Método Jacobi:", metodo_jacobi(A, b, [0, 0]))
print("Método Gauss-Seidel:", metodo_gauss_seidel(A, b, [0, 0]))
print("Método Richardson:", metodo_richardson(A, b, [0, 0]))


Método de Newton:
Iteración 1: x = [1.3333333333333333, 2.0]
Iteración 2: x = [1.2059047046763989, 1.7468850517832526]
Solución final: [1.2059047046763989, 1.7468850517832526]

Otros métodos numéricos:
Jacobi: [0.0, 0.0]
Gauss-Seidel: [0.0, 0.0]
Richardson: [0.3507672845301004, 1.0473362972982185]


### Conclusión



*   El método de Newton es el más efectivo en este caso, ya que converge rápidamente a una solución precisa en solo dos iteraciones. Proporciona la solución aproximada [1.2059047046763989, 1.7468850517832526].
*  El método de Jacobi no es adecuado para este sistema no lineal en particular, ya que no converge a una solución con los valores iniciales [1, 1]. En cambio, converge a la solución [0.0, 0.0]. Esto puede indicar que los valores iniciales están demasiado lejos de la solución o que el método de Jacobi no es apropiado para este sistema.
*   Al igual que el método de Jacobi, el método de Gauss-Seidel tampoco converge a una solución con los valores iniciales [1, 1]. En su lugar, converge a la solución [0.0, 0.0]. Esto sugiere que los valores iniciales están lejos de la solución o que el método de Gauss-Seidel no es apropiado para este sistema.
* El método de Richardson converge a una solución aproximada después de dos iteraciones. Proporciona la solución aproximada [0.3507672845301004, 1.0473362972982185].

Por lo cual, el método de Newton es el más eficiente y preciso para resolver este sistema no lineal. Los otros métodos numéricos, como Jacobi, Gauss-Seidel y Richardson, no logran converger a la solución correcta con los valores iniciales dados [1, 1].



